specs/trip_dispatch_spec.rb

```diff
@@ -111,4 +92,95 @@
+
+      expect(new_trip).must_be_kind_of RideShare::Trip
+    end
+
+    it "updates the passenger trip list" do
+      current_passenger = @dispatcher.find_passenger(3)
+      current_number_of_trips = current_passenger.trips.length
+      @dispatcher.request_trip(3)
+
+      expect(current_passenger.trips.length).must_equal (current_number_of_trips + 1)
+    end
+
+    it "updates the driver trip list" do
+      current_driver = @dispatcher.find_driver(5)
+      current_number_of_trips = current_driver.driven_trips.length
+      @dispatcher.request_trip(3)
+
+      expect(current_driver.driven_trips.length).must_equal (current_number_of_trips + 1)
+    end
+
+    it "ensures that driver assigned is available" do
+      driver = @dispatcher.check_driver_availability_and_assign(3)
+
+      expect(driver.status).must_equal :AVAILABLE
+    end
+
+    it "raises an ArgumentError if there are no available drivers" do
+      @dispatcher.request_trip(6)
+      @dispatcher.request_trip(3)
+
+      expect{@dispatcher.request_trip(1)}.must_raise ArgumentError
+    end
+
+    it "checks that drivers are not driving themselves" do
+      new_trip = @dispatcher.request_trip(5)
+      passenger = new_trip.passenger.id
+      driver = new_trip.driver.id
+
+      expect(passenger).wont_equal driver
+    end
+
+    it "will ensure that the driver who is selected has no trips" do
+      dispatcher_2 = RideShare::TripDispatcher.new(
+                                                  'specs/test_data/users_test.csv'
+                                                  'specs/test_data/trips_dispatcher_test.csv',
+                                                  'specs/test_data/driver_test_new.csv')
+      new_trip = dispatcher2.request_trip(3)
+      driver = new_trip.driver
+
+      expect(driver.trips.length).must_equal 1
+      expect(driver.id).must_equal 8
+    end
+
+  end
+
+  describe "In Progress Trips" do
+    it "will calculate total money spent even with an in progress trip" do
+      new_trip = @dispatcher.request_trip(3)
+      total_spent = new_trip.passenger.net_expenditures
+
+      expect(total_spent).must_equal 7
+    end
+
+    it "will calculate the average hourly revenue for a driver with an in progress trip" do
+      new_trip = @dispatcher.request_trip(3)
+
+      total_seconds_driven = 0
+      new_trip.driver.driven_trips.each do |trip|
+        if trip.end_time != nil
+          total_seconds_driven += trip.end_time - trip.start_time
+        end
+      end
+
+      total_hours_driven = (total_seconds_driven.to_f / 3600).round(2)
+      average_hourly_rate = (new_trip.driver.total_revenue / total_hours_driven).round(2)
+
+      expect(average_hourly_rate).must_equal 31.28
+    end
+
+    it "will ensure that the driver who is selected has no trips" do
+      new_trip = @dispatcher_2.request_trip(3)
+      driver = new_trip.driver
+
+      expect(driver.driven_trips.length).must_equal 1
+      expect(driver.id).must_equal 8
+    end
```

Comment:
What if there is no available driver? This is an important case, and is exactly the sort of situation where software will do something unexpected. You might set up the test as follows:

```ruby
it 'raises an error if no driver is available' do
  @dispatcher.drivers.each do |driver|
    # This may require an attr_writer in Driver
    driver.status = :UNAVAILABLE
  end
  proc {
    @dispatcher.request_trip(2)
  }.must_raise StandardError
end
```